### PR TITLE
IBX-1392: Fixed listening for encoded URI in clustered setup

### PR DESCRIPTION
**Pull request by barw4**
| Question                                  | Answer
| ---------------------------------------- | ------------------
| **JIRA issue**                          | [IBX-1392](https://issues.ibexa.co/browse/IBX-1392)
| **Type**                                   | bug
| **Target Ibexa version** | `v3.3`
| **BC breaks**                          | yes

As the title suggests. 

#### Checklist:
- [x] Provided PR description.
- [x] Tested the solution manually.
- [x] Provided automated test coverage.
- [x] Checked that target branch is set correctly (master for features, the oldest supported for bugs).
- [x] Ran PHP CS Fixer for new PHP code (use `$ composer fix-cs`).
- [x] Asked for a review (ping `@ezsystems/engineering-team`).

### CODE REVIEW

File: eZ/Bundle/EzPublishIOBundle/EventListener/StreamFileListener.php [hunk 1]
```diff
@@ -62,7 +62,11 @@ public function onKernelRequest(RequestEvent $event)
 
         $binaryFile = $this->ioService->loadBinaryFileByUri($uri);
         if ($binaryFile instanceof MissingBinaryFile) {
-            throw new NotFoundHttpException("Could not find 'BinaryFile' with identifier '$uri'");
+            // Try again with decoded URI
+            $binaryFile = $this->ioService->loadBinaryFileByUri(urldecode($uri));
+            if ($binaryFile instanceof MissingBinaryFile) {
+                throw new NotFoundHttpException("Could not find 'BinaryFile' with identifier '$uri'");
+            }
```
Comment (adamwojs) [inline eZ/Bundle/EzPublishIOBundle/EventListener/StreamFileListener.php]: Q: What will happen if my file is named `obraze%20z spacja.jpg`?

**Review by adamwojs — CHANGES_REQUESTED** [review]
"Guessing" correct file name it's an workaround, not proper bugfix. 

IMHO the source of the regression is https://github.com/ezsystems/ezpublish-kernel/pull/3126 and this patch should be reverted (on v2.5 as well).

File: eZ/Bundle/EzPublishIOBundle/EventListener/StreamFileListener.php (hunk 1 above)
Comment (glye) [reply eZ/Bundle/EzPublishIOBundle/EventListener/StreamFileListener.php]: The proper solution for all these troubles is to rename ALL uploaded files to random hashes. Problem solved, and it has other benefits (unguessable). But that's a bigger change. So do we want a quick solution/workaround now or not? A real fix should come later in any case.

Comment (mateuszbieniek) [reply eZ/Bundle/EzPublishIOBundle/EventListener/StreamFileListener.php]: @adamwojs https://github.com/ezsystems/ezpublish-kernel/pull/3126 was fixing a different issue, and with or without these changes ibx-1392 is still valid.

Comment (barw4) [reply eZ/Bundle/EzPublishIOBundle/EventListener/StreamFileListener.php]: Both @adamwojs and @mateuszbieniek are right, to some extent. Mateusz's PR fixed the following problematic filenames:
`some%25special+chars.jpg`. 
However, https://github.com/ezsystems/ezpublish-kernel/pull/3126 broke the following:
`test test.png`.

The current PR handles both these cases,  we can also go as a temporary solution with just:
`        $binaryFile = $this->ioService->loadBinaryFileByUri(urldecode($uri));` - that would mean just a single load. Obviously, this isn't ideal but handles most of the cases.

Without guessing: https://github.com/ezsystems/ezplatform-kernel/pull/272/commits/610e6f02598ce67c4e1ef49aca8a5ead1148c94e

Comment (barw4) [reply eZ/Bundle/EzPublishIOBundle/EventListener/StreamFileListener.php]: > Q: What will happen if my file is named `obraze%20z spacja.jpg`?

@adamwojs This will not work in any of those PRs or even before those PRs were introduced.

Comment (alongosz) [reply eZ/Bundle/EzPublishIOBundle/EventListener/StreamFileListener.php]: > > Q: What will happen if my file is named `obraze%20z spacja.jpg`?
> 
> This will not work in any of those PRs or even before those PRs were introduced.

But do we have a solution for such case?

Comment (barw4) [reply eZ/Bundle/EzPublishIOBundle/EventListener/StreamFileListener.php]: Besides random hashes, I don't know tbh. The problem is that the browser converts the space to `%20` before a request is made so we cannot guess which `%20` is a space or which is just `%20`. 

However, this seems like an edge case, there will always be some names that won't be parsed properly like `test$%^&_ds.jpg`. This file for example breaks on the webserver level, even before reaching PHP. Until the proper solution, which probably utilizes hashes, is made there will be always some problematic filenames.

Comment (mateuszbieniek) [reply eZ/Bundle/EzPublishIOBundle/EventListener/StreamFileListener.php]: @glye I agree but we can't apply this solution here. In most cases images are served directly from the webserver, therefore introducing hashing will either force everyone to stream files via index.php, or at least will require a change in generated URLs, which I think both are out of the question.

Comment (glye) [reply eZ/Bundle/EzPublishIOBundle/EventListener/StreamFileListener.php]: (Continuing the renaming/streaming discussion elsewhere)

Comment (YohannsMonnier) [reply eZ/Bundle/EzPublishIOBundle/EventListener/StreamFileListener.php]: > bigger

I think we should only sluggify it instead of changing it to a random name, for SEO perpective having name with keyword in it could be a good point.

Yohann

**Review by alongosz — APPROVED** [review]

**Comment by YohannsMonnier** [comment]
@barw4 Do you have a date in mind for the final fix ? 

We have patched our installation, by modifying the vendor manually, but we really would like to have a real fix.

Thank you very much, 
Yohann

**Comment by barw4** [comment]
> @barw4 Do you have a date in mind for the final fix ?
> 
> We have patched our installation, by modifying the vendor manually, but we really would like to have a real fix.
> 
> Thank you very much, Yohann

Hi Yohann, can we stick to the ticket, please? I have responded there to you.

File: eZ/Bundle/EzPublishIOBundle/Tests/EventListener/StreamFileListenerTest.php [hunk 2]
```diff
@@ -97,29 +97,39 @@ public function testRespondsToIoUri()
...
+
+        $finalUri = sprintf('http://%s%s', $host, $uri);
 
         $this->ioServiceMock
-            ->expects($this->once())
+            ->expects(self::once())
             ->method('loadBinaryFileByUri')
-            ->with(sprintf('http://%s%s', $host, $uri))
-            ->will($this->returnValue($binaryFile));
+            ->with(urldecode($finalUri))
```
Comment (Steveb-p) [inline eZ/Bundle/EzPublishIOBundle/Tests/EventListener/StreamFileListenerTest.php:122]: Do not use the decoding mechanism as the implementation uses. Instead, use raw value, so we can be certain about inputs and outputs, and that we do not encode / decode twice by mistake.

**Review by Steveb-p — CHANGES_REQUESTED** [review]

**Comment by barw4** [comment]
Closed as another solution was chosen.
